stan/math/prim/meta/is_matrix.hpp

```diff
@@ -0,0 +1,23 @@
+#ifndef STAN_MATH_PRIM_META_IS_MATRIX
+#define STAN_MATH_PRIM_META_IS_MATRIX
+
+#include <stan/math/prim/meta/disjunction.hpp>
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/meta/is_var.hpp>
+#include <stan/math/prim/meta/is_rev_matrix.hpp>
+
+namespace stan {
+/**
+ * Check if a type is derived from `Eigen::EigenBase` or is a `var_value`
+ *  whose `value_type` is derived from `Eigen::EigenBase`
+ * @tparam T type to check.
+ * @ingroup type_trait
+ */
+template <typename T>
+struct is_matrix
```

Comment:
I like `is_matrix_like`! @t4c1 do you have an opinion here? Also @t4c1 one side question, should `is_matrix_like` also support `matrix_cl<>` objects (aka should multiply support `var_value<matrix_cl<double>>`? I think it should work right? Or should we wait on that?

Reply:
I prefere `is_matrix` or maybe `is_stan_matrix`. Regardless of what we decide on, it will be confusing as this accepts all matrices, while `is_eigen_matrix` only accepts non-vector matrices (to be clear in my opinion `is_eigen_matrix` is the one that should be changed).

If you make the same code work for `matrix_cl`s that is great! However, we will need some requires to differentiate between Eigen types and matrix_cl, as underlying systems (kernel generator and Eigen) are different enough that we will not always be able to use the same code for both implementations.

Reply:
> I prefere `is_matrix` or maybe `is_stan_matrix`. Regardless of what we decide on, it will be confusing as this accepts all matrices, while `is_eigen_matrix` only accepts non-vector matrices (to be clear in my opinion `is_eigen_matrix` is the one that should be changed).

I like the idea of changing `is_eigen_matrix` to `is_eigen_dynamic_matrix` since we check that rows and cols at compile time are -1.

Making it `is_matrix_like` also means that `is_vector_like` needs to change

> If you make the same code work for matrix_cls that is great! However, we will need some requires to differentiate between Eigen types and matrix_cl, as underlying systems (kernel generator and Eigen) are different enough that we will not always be able to use the same code for both implementations.

if we apply the right tlc then I think it's possible, the only thing I need to think about here is when one is an opencl matrix and the other is an eigen matrix.

Reply:
> changing is_eigen_matrix to is_eigen_dynamic_matrix

Seems good.

I am happy with `is_matrix`, `is_stan_matrix`, or `is_matrix_like`.

Reply:
Alright so I changed `is_eigen_matrix` to `is_eigen_matrix_dynamic`, then we will have `is_matrix` for this. I also changed the definition of `is_vector` because if we have `is_matrix` I think we need the same thing for `is_vector`.

@t4c1 I think this could work for both eigen and `matrix_cl` types but we should do it in a seperate PR. We need to add methods to `matrix_cl` for `dot()`, `transpose()`, `array()` (and I think we need a += operator as well). Then we also need to add `matrix_cl<>` to the test suite so we can check gradients and stuff

Reply:
Agreed on separate PR.